연락처관리 UI [STEP 3] Mason, D.O

doshkor opened the pull request:
안녕하세요 제이슨! @ehgud0670

연락처 관리 UI 버전 STEP3 PR 요청드립니다 :)

이번 프로젝트의 중점 학습 내용은 델리게이트 패턴, TextField 였습니다 :)

전체적으로 네이밍이나 구조가 어색하다고 느껴지시는 부분은 얼마든지 짚어주시면 감사하겠습니다!

# ✅ 요구사항, 구현 내용

- [x]  연락처 목록 화면의 우상단 버튼을 통해 연락처 추가 화면에 진입할 수 있습니다
- [x]  오토 레이아웃과 스택뷰를 활용하여 화면을 레이아웃합니다
- [x]  각 필드에 맞는 키보드 종류를 지정합니다
- [x]  취소 버튼을 선택하면 정말 취소할 것인지 묻도록합니다
- [x]  저장 버튼을 선택하면 입력한 정보가 올바르게 입력되었는지 확인하도록 합니다
    - [x]  이름은 사용자가 중간에 띄어쓰기를 하더라도 띄어쓰기를 제거합니다
    - [x]  나이는 숫자로만 입력해야하며, 세 자리수 이하여야 합니다.
    - [x]  연락처는 중간에  `-`로 구분하며, `-`는 두 개 존재해야 합니다.
        - [x]  을 제외하고 숫자는 9자리 이상이어야 합니다.
- [x]  모든 정보가 올바로 입력되었다면 저장 버튼을 선택했을 때 이전 화면으로 돌아가고, 추가한 연락처가 연락처 리스트에 추가됩니다

# ❓ 트러블슈팅

### 1. textField.text 에 format(##-###-###) 입히기

유저의 입력을 Character로 받는 `textField(shouldChangeCharactersIn …)-> Bool` 함수안에서 유저가 숫자 입력시 textField.text를 연락처형식(##-###-###)으로 바꿔주는 로직을 의도하였는데요. 이 함수는 유저가 입력한 `Character`를 받을지 말지 `Bool` 타입으로 주는 함수로, true 를 리턴하면 함수가 종료 후에야 `Character`가 추가가 되는 문제가 발생하였습니다. 

즉, 문자가 추가된 값에 대하여 format을 입혀야 하는데, 이 함수안에 로직을 작성하게되면 format을 입힌 값에 대하여 문자가 추가되는 문제가 발생하였어요.

그래서 `didSet{ }` 과 같은 방식을 사용하면 좋겠다고 생각하였는데요. 그렇게 되면 입력받은 문자는 그대로 전달해주고, 보여지는 부분에서 format을 입혀 보여주기만 하면 되서 로직이 분리가 될거라고 생각했어요.

먼저 IBOutlet 변수는 UITextField! 타입이고, 저희가 원하는 text 속성에 대한 didSet 접근이 어렵다고 생각하였어요. 그래서 format을 입히기 원하는 textField에 `addTarget(… for: .editingChanged)` 하는 방식으로 구현해보았어요

**[ 기존 코드 ]**

```swift
func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
		if textField == self.phoneNumberTextField {
		    var currentText = textField.text ?? ""
		    if string.isEmpty {
		        currentText.removeLast()
		    }

		    let newText = currentText + string
		    textField.text = newText.formattedPhoneNumber
		
		    return false  
		}
}
```

- 앞서 말씀드린 오류를 수정하기 위해서 문자를 미리 추가( `+ string` )하고 text 에 format 을 입힌 값을 text에 넣어주고, 입력받은 문자가 추가되는 것을 막는 방식(`return false`)으로 구현하였어요.

**[ 수정 코드 ] - addTarget** 

```swift
phoneNumberTextField.addTarget(self, action: #selector(self.didPhoneNumberTextChange), for: .editingChanged)

@objc func didPhoneNumberTextChange() {
        phoneNumberTextField.text = phoneNumberTextField.text?.formattedPhoneNumber
}

func textField(_ textField: UITextField,
                   shouldChangeCharactersIn range: NSRange,
                   replacementString string: String) -> Bool {
        if textField == ageTextField {
            guard string.isEmpty || Int(string) != nil else {
                return false
            }
        }
        return true
    }
```

- `addTarget` 으로 구현하여 `textField(shouldChangeCharactersIn …)-> Bool` 함수에서는 숫자가 아닌 문자를 입력을 방지하는 코드만 작성하면 되게 되었어요

❓😎 이렇게 addTarget 을 줘서 입력을 제한하는 것과 같은 효과를 부여하는 방식도 적절한지요?

### 2. Custom Class 로 구현

위처럼 구현한 로직을 필요하면 재사용할 수 있도록, `PhoneNumberText` 라는 `Custom Class`로 분리하였습니다.

해당 `TextField`가 `UITextFieldDelegate` 를 구현하고 스스로의 delegate가 되어, 기존에 해 주었던 로직들을 해당 커스텀 클래스 내부로 옮겼습니다.

또한, 전화번호 형식으로 입력 text를 Formatting 하는 String Extension 들은 오직 전화번호 입력 텍스트 필드 내부에서만 쓰일 메서드라고 생각되어 해당 로직들 또한 커스텀 클래스가 갖게끔 구현하였습니다.

❓😎 이렇게 커스텀 클래스로 분리한 것이 적절 해 보이는지요?

여기서 저희가 발견한 문제는… 만약 누군가가 기본 구현 된 해당 커스텀 텍스트 필드의 delegate 를 따로 지정 해 주게 된다면 기존에 지정한 delegate 가 해제되기에 저희가 구현 해 놓은 로직이 사라진다는 문제가 발생하였습니다.

> 이를 해결하기 위해… 
1. delegate 배열을 만들어 해당 delegate 들이 모두 호출되게 하는 방식이나…
2. override delegate 의 set 메서드를 부여해 새로운 delegate 할당을 막는다거나… 
3. Protocol 의 기본 구현… 
등등의 여러 방식을 시도 해 보았지만 해결하지 못했습니다😭
> 

❓🥲 애초에 접근이 잘못된 것이라는 생각도 드는데요, 혹시 이러한 문제를 제이슨은 어떻게 해결하실 것 같으신지 여쭤보고 싶습니다!

또한 저희 프로젝트는 스토리보드 방식이었기에 init(coder:) 에서 초기 세팅을 해 줘야 했지만, Code UI로 생성 하게 될 때를 대비해서 초기 세팅을 init(frame:) 에서도 호출 해 주었습니다. 
이렇게 각각의 initializer 에서 초기화 해 주는 방식이 적절할까요?

altokeys committed:
fix: reuseIdentifier 리턴값 수정

qwerty3345 committed:
feat: 연락처 리스트 화면 네비게이션바 추가

altokeys committed:
fix: Navigation Controller 방식으로 수정

altokeys committed:
feat: 연락처 추가 화면 추가

qwerty3345 committed:
feat: 연락처 추가 화면 취소 버튼 기능 구현

- 유저에게 얼럿을 띄우고 정말로 취소할지 확인하게 함

altokeys committed:
feat: save 버튼 기능구현

qwerty3345 committed:
feat: 연락처 추가 시 테이블뷰 리로드 구현

- Delegate 패턴을 활용하여 알려줌

altokeys committed:
feat: 나이입력시 문자입력방지 기능구현

qwerty3345 committed:
refactor: private 키워드 은닉화

qwerty3345 committed:
feat: 사용자 입력 검증 및 알림 로직 구현

- 연락처 정보로 변환하지 못할 경우, 해당 실패 케이스에 따라 유저에게 얼럿을 띄워줌

altokeys committed:
feat: 전화번호 입력시 하이픈(-) 자동입력 기능 구현

qwerty3345 committed:
refactor: 함수 분리, 불필요 주석 및 개행 제거

qwerty3345 committed:
refactor: 연락처 입력 TextField Custom Class 로 변경

qwerty3345 committed:
refactor: 전화번호 Formatting 관련 String Extension 메서드 TextField 커스텀 클래스로 이동

altokeys committed:
refactor: magic string 상수화

ehgud0670 commented on ContactManager/ContactManagerUI/Extensions/Extension+UITableViewCell.swift:
```diff
@@ -13,7 +13,7 @@ protocol ReusableView {
 extension ReusableView {
     static var reuseIdentifier: String {
-        String(describing: ContactCell.self)
+        String(describing: self)
```
굿입니다 👍

ehgud0670 commented on ContactManager/ContactManagerUI/View/PhoneNumberTextField.swift:
```diff
@@ -0,0 +1,91 @@
+            return result
+        }
+
+        return result
```
pattern을 let 변수로 바꾸고 pattern 먼저 구한 다음에 applyPatternOn을 호출해서 result 를 반환해도 될 것 같아요. 
음,, 지금은 처음 봤을 때 가독성이 좋지 못해 보입니다. 🤔

qwerty3345 replied:
로직을 분리하였습니다!
> 해당 커밋: ddee2e1

ehgud0670 commented on ContactManager/ContactManagerUI/View/PhoneNumberTextField.swift:
```diff
@@ -0,0 +1,91 @@
+
+// MARK: - Number Format 관련
+extension PhoneNumberTextField {
+    func formattedPhoneNumber(with text: String?) -> String {
```
모델 값을 뷰에게 보일수 있도록 값을 정제해서 반환하는 역할은 뷰가 아닌 Presenter(패턴은 MVP)나 ViewModel(패턴은 MVVM) 객체가 해야 합니다.
PhoneNumber 만이라도 formattedPhoneNumber 와 applyPatternOn 메소드를 뷰가 가지지 않도록 Presenter나 ViewModel를 만들어 역할을 분리해보세요(MVC 패턴을 지향한다고 하면 뷰컨트롤러가 해당 메소드를 갖게 하세요).

qwerty3345 replied:
MVC 구조로 구현하였고, 값을 정제해서 변환하는 역할을 뷰컨트롤러가 수행하게끔 로직을 변경하였습니다.
> 해당 커밋: d593c7d

ehgud0670 commented on ContactManager/ContactManagerUI/ViewController/AddContactViewController.swift:
```diff
@@ -0,0 +1,92 @@
+        }
+        
+        dismiss(animated: true)
+        delegate?.didContactChanged()
```
catch 문에서 return 문이 없도록 만들수 있지 않을까요?

qwerty3345 replied:
return 문이 없도록 수정하였습니다!
> 해당 커밋: 5bd7e85

ehgud0670 commented on ContactManager/ContactManagerUI/ViewController/AddContactViewController.swift:
```diff
@@ -0,0 +1,92 @@
+
+final class AddContactViewController: UIViewController {
+    // MARK: - Properties
+    var delegate: AddContactViewControllerDelegate?
```
```suggestion
    weak var delegate: AddContactViewControllerDelegate?
```

doshkor replied:
delegate 패턴을 사용하면 위임을 하는 인스턴스와 delegate의 인스턴스가 서로를 강하게 가르키게 된다는 것을 알게되었습니다.!
때문에 weak 키워드를 붙이고, 프로토콜에 클래스에서만 사용하도록 `:AnyObject` 를 추가하였습니다

ehgud0670 commented on ContactManager/ContactManagerUI/ViewController/AddContactViewController.swift:
```diff
@@ -0,0 +1,92 @@
+import UIKit
+
+protocol AddContactViewControllerDelegate {
+    func didContactChanged()
```
델리게이트 메소드는 해당 스타일을 참고하시면 좋아요.

```swift
func namePickerView(_ namePickerView: NamePickerView, didSelectName name: String)
func namePickerViewShouldReload(_ namePickerView: NamePickerView) -> Bool
```

link: https://github.com/kodecocodes/swift-style-guide#delegates

ehgud0670 replied:
그리고 왜 이렇게 스타일을 고수해야 하는지도 고민해보세요.

doshkor replied:
깃허브 스타일 가이드를 참고하면서 mason과 이야기를 해보았는데요.!
위임을 하는 클래스나 위임을 받은 클래스에서 함수명을 보고 명시적으로 어느 클래스에서 위임한 함수인지 그리고 어떤 역할을 하는지를 알 수 있을 것이고, 파라미터 또한 같은 의미로 볼 수 있다고 생각했습니다.! 🙆‍♂️

ehgud0670 commented on ContactManager/ContactManagerUI/ViewController/AddContactViewController.swift:
```diff
@@ -0,0 +1,92 @@
+        }
+        
+        dismiss(animated: true)
+        delegate?.didContactChanged()
```
dismiss 하고나서 메소드 호출을 하는 것이 목적이라면 completion을 써도 좋을 것 같아요. dismiss 후에 호출되는 것이 100% 보장되니까요. 

```suggestion
        dismiss(animated: true, completion: {
            self.delegate?.didContactChanged()
        })
```

qwerty3345 replied:
dismiss 에도 completion 이 있다는 사실을 덕분에 알게 되었습니다 :)
dismiss 이후 동작 수행 시 메모리에서 해제되지 않으리라는 보장이 없다는 부분을 덕분에 생각하게 되었습니다.
감사합니다!
> 해당 커밋: 5bd7e85

ehgud0670 commented on ContactManager/ContactManagerUI/ViewController/ContactViewController.swift:
```diff
@@ -33,3 +39,9 @@ final class ContactViewController: UIViewController {
+
+extension ContactViewController: AddContactViewControllerDelegate {
+    func didContactChanged() {
+        contactTableView.reloadData()
```
1) 이전 미션에서 모델 객체를 싱글톤으로 써보셨으니 이번에는 쓰지 말고 뷰컨트롤러끼리 직접 모델값을 주고 받도록 해보세요. 
2) 1을 구현한뒤 didContactChanged 메소드에는 AddContactVC에서 전달된 모델값만 ContactVC의 모델객체에 add 하는 코드만 두세요.
3) 2를 구현했다면 모델객체가 노티(Notification)를 통해 모델값이 변했다는 것을 ContactVC가 알리고 해당 노티를 받은 ContactVC가 테이블뷰를 reloadData를 호출하도록 만들어보세요. 3까지 가능하다면 결국에 ContactViewController.swift 34줄 fetchContactData 메소드에서 뷰를 업데이팅하는 코드(`contactTableView.reloadData()`)를 분리할 수 있게 됩니다. 

이러한 주문을 한 이유는 기본적인 MVC 패턴을 수행하시라고 주문을 드렸습니다. 그럼 화이팅입니다!

doshkor replied:
1. 싱글톤을 사용하지 않고 데이터를 주고 받도록 구현해보았습니다.
2. 저장 버튼을 누르면 delegate를 통해 contact를 ContactVC의 contacts에 append 합니다
3. ContactVC의 contacts에 didSet으로  ContactDatasource.contacts 값을 변경하고 reload()합니다

해당커밋: 51dc617

ehgud0670 reviewed (changes requested):
## 질문에 대한 답변!

1) `addTarget`
=> 네 해주셔도 됩니다!
2) 해당 커스텀 클래스가 델리게이트를 이미 할당했는데 이후에 다른 객체도 delegate 역할을 할당해 역할을 뺏어버리는 버그는 굳이 커스텀 클래스가 아니어도 가능한 실수이기 때문에 다른 방법은 생각이 안납니다! 
최대한 동료들에게 해당 커스텀 클래스의 사용 용도를 알려주는게 맞지 않나 생각합니다.
3) 네, 둘 다 대비하면 더 좋습니다 🙂

제가 여러 패턴(MVC, MVP, MVVM)을 언급했는데 먼저 MVC 부터 하시길 권장합니다.
그리고 모르시는거 있으면 언제든지 질문주세요~

altokeys committed:
refactor: formattedPhoneNumber 구현로직 수정

qwerty3345 committed:
refactor: TextField View가 데이터를 수정하지 않도록 MVC 구조로 변경

altokeys committed:
refactor: return문 제거 및 completion 추가

altokeys committed:
refactor: delegate 컨벤션 적용

altokeys committed:
refactor: 데이터 전달 방식 변경

qwerty3345 committed:
chore: 개행 제거, MARK 주석 추가

ehgud0670 commented on ContactManager/ContactManagerUI/ViewController/ContactViewController.swift:
```diff
@@ -10,26 +10,44 @@ import UIKit
+    
     private let contactDataSource = ContactDataSource()
-
+    private var contacts: [Contact] = [] {
```
contacts 모델 값은 오직 ContactDataSource 만 가지도록 다시 리팩토링 해보세요.
그리고 노티를 이용해 모델값의 변화가 있을때 reloadData를 하도록 리팩토링 해보세요 :)

doshkor replied:
ContactDataSource의 contact에 NotificationCenter 을 추가하여 데이터 변경시 post 하고
ContactViewController 에서 옵저버 추가하여 reload() 하게 변경해보았습니다.!

해당 커밋: f6eb6cb

ehgud0670 replied:
잘 변경하셨습니다! 👍

altokeys committed:
refactor: contact 추가시 notification 적용

qwerty3345 committed:
chore: 사소한 변경

- Notification.Name 을 NSNotification.Name 으로 변경 (typealias로 동일하지만, 컴파일러가 가끔 추론을 못해서 변경)
- 메서드 위치 조절

ehgud0670 reviewed (approved):
고생하셨습니다 :)
